R/featurization/get_features.R

```diff
@@ -0,0 +1,250 @@
+
+
+#
+# Featurization script
+# 
+
+setwd("C:/users/ben_ryan/documents/git/ugb/R/featurization")
+source("clustering.R")
+source('glove_analysis.R')
+source('bias_themes_analysis.R')
+
+registerDoParallel(detectCores())
+
+#
+# Get the raw text data for embedding training
+# 
+
+b_corp <- fread('C:/users/ben_ryan/documents/darpa ugb/breitbart_texts.csv',
+                header=FALSE)
+
+bv = trainEmbeddings(b_corp$V2, alpha=0.05)
+write.table(bv,
+            'breitbart_embeddings_wComments.txt',
+            col.names=FALSE,
+            quote=FALSE,
+            fileEncoding='utf-8')
+
+# To read in again from saved:
+# bv = fread('breitbart_embeddings_wComments.txt', header=FALSE)
+# terms <- bv$V1
+# bv[,V1:=NULL]
+# bv <- data.matrix(bv)
+# rownames(bv) <- terms
+
+#
+#
+# Begin feature extraction from articles and comments datasets
+#
+# 
+
+barts = fread('breitbart_articles.csv') #load breitbart articles
+bcs = fread('breitbart_comments.csv') #load breitbart comments
+
+# 
+# Derive bias component vectors
+# 
+
+gender_bias <- deriveBias(genderPairs, wv=bv, method='pca', diag=TRUE)
+race_bias <- deriveBias(adj_race_names, wv=bv, diag=TRUE)
+power_bias <- deriveBias(powerPairs, wv=bv, method='pca', diag=TRUE)
+
+#
+# Derive embeddings for articles and article titles, and merge into one 
+# data.table. 
+# 
+
+arts_text_tokens = getTokens(barts$art_text)
+bgloved_arts <- lapply(arts_text_tokens, gloved, wv=bv)
+bgloved_arts <- data.table(do.call(rbind, bgloved_arts))
+
+# 
+# Propagate bias components
+#
+
+gender <- foreach(i = 1:nrow(bgloved_arts),
+                  .inorder=TRUE,
+                  .export=c('biasComponent','avgBias','bv','gender_bias','power_bias','race_bias'),
+                  .packages = c('NLP','openNLP','dplyr'),
+                  .multicombine = TRUE, 
+                  .options.multicore = list(preschedule = FALSE)) %dopar% {
+            avgBias(arts_text_tokens[[i]], gender_bias, bv)
+        }
+
+bgloved_arts$gender <- unlist(gender)
+
+
+race <- foreach(i = 1:nrow(bgloved_arts),
+                  .inorder=TRUE,
+                  .multicombine = TRUE, 
+                  .options.multicore = list(preschedule = FALSE)) %dopar% {
+            avgBias(arts_text_tokens[[i]], race_bias, bv)
+        }
+
+bgloved_arts$race <- unlist(race)
+
+
+power <- foreach(i = 1:nrow(bgloved_arts),
+                  .inorder=TRUE,
+                  .multicombine = TRUE, 
+                  .options.multicore = list(preschedule = FALSE)) %dopar% {
+            avgBias(arts_text_tokens[[i]], power_bias, bv)
+        }
+
+bgloved_arts$power <- unlist(power)
+
+# 
+# Set column names and add article id
+#
+
+names(bgloved_arts) <- paste('art_text_embeddings_',names(bgloved_arts), sep='')
```

Comment:
FYI, if `sep = ''` in the `paste()` function, you can just use `paste0()` within an explicit `sep` parameter, i.e., `sep` is assumed to be `sep = ''`.